pkg/services/authn/grpcutils/config.go

```diff
@@ -0,0 +1,68 @@
+package grpcutils
+
+import (
+	"fmt"
+
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type Mode string
+
+func (s Mode) IsValid() bool {
+	switch s {
+	case ModeOnPrem, ModeCloud:
+		return true
+	}
+	return false
+}
+
+const (
+	ModeOnPrem Mode = "on-prem"
+	ModeCloud  Mode = "cloud"
+)
+
+type GrpcClientConfig struct {
+	Token            string
+	TokenExchangeURL string
+	TokenNamespace   string
+	Mode             Mode
+}
+
+func ReadGrpcClientConfig(cfg *setting.Cfg) (*GrpcClientConfig, error) {
+	section := cfg.SectionWithEnvOverrides("grpc_client_authentication")
+
+	mode := Mode(section.Key("mode").MustString(string(ModeOnPrem)))
+	if !mode.IsValid() {
+		return nil, fmt.Errorf("grpc_client_authentication: invalid mode %q", mode)
+	}
+
+	return &GrpcClientConfig{
+		Token:            section.Key("token").MustString(""),
+		TokenExchangeURL: section.Key("token_exchange_url").MustString(""),
+		TokenNamespace:   section.Key("token_namespace").MustString("stacks-" + cfg.StackID),
+		Mode:             mode,
```

Comment:
do we need mode?   is that the same as having a "remote_address" or a token_exchange_url that is not empty?

(weirdly this would be the first place we have a setting that explicitly says onprem vs cloud)

Reply:
Unfortunately we do, because for both cloud and on-prem setups the config options will have non-empty values. Now, we could potentially identify the type of setup based on the values of `remote_address` and `TokenExchangeURL`, but I feel that's hacky and non-obvious.

However, this will also go away as soon as we support Access tokens on-prem. This is really the main difference between the two setups.

Reply:
if it is just a temporary thing and only in cloud -- you can use the existence of `cfg.StackID` as an indication we are in "cloud mode" rather than adding a new setting

Reply:
hmm, yes, on client-side we can make use of `cfg.StackID`, if it's guaranteed that it's going to be empty for on-prem. Is this the case?

Reply:
@gamab, iirc, we debated this some time ago. Do you remember what the arguments against using `cfg.StackID` were?